pkg/ottl/functions.go

```diff
@@ -15,12 +15,35 @@
 package ottl // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
 
 import (
+	"context"
 	"errors"
 	"fmt"
 	"reflect"
+	"strconv"
 	"strings"
+
+	"go.opentelemetry.io/collector/component"
 )
 
+// Arguments holds the arguments for an OTTL function, with arguments
+// specified as fields on a struct. Argument ordering is defined
+type Arguments interface{}
+
+// Factory defines an OTTL function factory that will generate an OTTL
+// function to be called based on an invocation within a statement.
+type Factory[K any] interface {
+	// Name is the canonical name to be used by the user when invocating
+	// the function generated by this Factory.
+	Name() string
+
+	// CreateEmptyArguments initializes an Arguments struct specific to this
+	// Factory containing the arguments for the function.
+	CreateEmptyArguments() Arguments
+
+	// CreateFunction creates an OTTL function that will use the given Arguments.
+	CreateFunction(ctx context.Context, set component.TelemetrySettings, args Arguments) (ExprFunc[K], error)
```

Comment:
@evan-bradley good call not including `ctx`.

Reply:
This was a concern I had as well. According to the [Go language developers](https://go.dev/blog/context-and-structs), we shouldn't add `context.Context` to any structs, which seems to be unconventional anyway, so I didn't add it to the struct. As a result, the existing struct feels a little strange since it only has one field. I've pushed a commit that includes that change.